### PR TITLE
Yoshi serve command

### PR DESCRIPTION
**Pull request by Gongreg**
Issue: #914.

### 🔦 Summary

Added serve command to yoshi-flow-app. It runs dev server in production mode and also starts cdn to serve statics created with yoshi build.

### 🗺️ Test Plan

Added feature tests for yoshi serve (to check if it starts and if it properly checks for missing build).

Added unit tests for packageJSON scripts parser (yoshi serve uses same entry point as the one it finds in yoshi start).

### CODE REVIEW

File: packages/yoshi-flow-app/src/scripts/serve.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+  const serverProcess = await ServerProcess.create({
+    serverFilePath,
+    suricate: config.suricate,
+    appName: config.name,
+  });
+
+  await Promise.all([serverProcess.initialize(), startCDN(config)]);
+
+  const actualStartUrl = config.suricate
```
Comment (Gongreg) [inline packages/yoshi-flow-app/src/scripts/serve.ts]: Duplicated code from devEnvironment.

File: packages/yoshi-flow-app/src/scripts/serve.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        --debug         Allow app-server debugging
+        --debug-brk     Allow app-server debugging, process won't start until debugger will be attached
+        --url           Opens the browser with the supplied URL
+    `,
+    );
+
+    process.exit(0);
+  }
+
+  console.log(chalk.cyan('Starting the environment...\n'));
```
Comment (Gongreg) [inline packages/yoshi-flow-app/src/scripts/serve.ts]: Maybe we could add a better message?

File: packages/yoshi-flow-app/src/bin/yoshi-app.ts [hunk 3]
```diff
@@ -75,20 +76,22 @@ Promise.resolve().then(async () => {
...
   await verifyDependencies();
 
+  const config = loadConfig();
+
   if (command === 'start') {
     process.env.NODE_ENV = 'development';
     process.env.BABEL_ENV = 'development';
+  }
 
+  if (command === 'start' || command === 'serve') {
```
Comment (Gongreg) [inline packages/yoshi-flow-app/src/bin/yoshi-app.ts]: I found it strange, but I couldn't set the env variables inside the command itself.

It seems like there is some code which is imported before running the command that depends on the variables.

I didn't investigate it too much, but I don't like that inner details of commands is leaking here.

File: packages/yoshi-flow-app/src/scripts/serve.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    console.log(
+      chalk.red(
+        `${config.clientFilesPath} is missing. Run \`yoshi build\` and try again.`,
+      ),
+    );
+
+    process.exit(1);
+  }
+
+  console.log(chalk.cyan('Starting the environment...\n'));
```
Comment (Gongreg) [inline packages/yoshi-flow-app/src/scripts/serve.ts]: A better message would be better

**Comment by ranyitz** [comment]
Is it still a work in progress? If no, let's change the title.

In any case It will take some time for me to go over it, so i'm sorry in advance 😭

File: packages/yoshi-flow-app/src/scripts/serve.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    appName: config.name,
+  });
+
+  await Promise.all([serverProcess.initialize(), startCDN(config)]);
+
+  const actualStartUrl = config.suricate
+    ? getTunnelUrl(config.name)
+    : url || 'http://localhost:3000';
+
+  openBrowser(actualStartUrl);
```
Comment (ranyitz) [inline packages/yoshi-flow-app/src/scripts/serve.ts]: remove open browser

File: packages/yoshi-flow-app/src/scripts/serve.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    suricate: config.suricate,
+    appName: config.name,
+  });
+
+  await Promise.all([serverProcess.initialize(), startCDN(config)]);
+
+  const actualStartUrl = config.suricate
+    ? getTunnelUrl(config.name)
+    : url || 'http://localhost:3000';
+
```
Comment (ranyitz) [inline packages/yoshi-flow-app/src/scripts/serve.ts]: Add log when the both are ready (cdn, app-server), something like.

Use the same messaging as the devEnvironment logger. - https://github.com/wix/yoshi/blob/master/packages/yoshi-common/src/dev-environment-logger.ts#L22

Separate for dev-server and app-server.

File: packages/yoshi-flow-app/src/scripts/serve.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        `${config.clientFilesPath} is missing. Run \`yoshi build\` and try again.`,
+      ),
+    );
+
+    process.exit(1);
+  }
+
+  console.log(chalk.cyan('Starting the environment...\n'));
+
+  const serverProcess = await ServerProcess.create({
```
Comment (ranyitz) [inline packages/yoshi-flow-app/src/scripts/serve.ts]: either implement `prodcutionServerProcess` or extend the server process to not run anything "dev" related (e.g. socket server for hmr)

File: packages/yoshi-common/src/server-process.ts [hunk 8]
```diff
@@ -91,46 +81,93 @@ export default class ServerProcess {
...
   }
 
-  onMessage(response: any) {
-    this.resolve && this.resolve(response);
+  async close() {
+    // @ts-ignore
+    if (this.child && this.child.exitCode === null) {
+      this.child.kill();
+
+      await new Promise(resolve => {
```
Comment (ranyitz) [inline packages/yoshi-common/src/server-process.ts]: Nice catch!

File: packages/yoshi-common/src/server-process.ts [hunk 9]
```diff
@@ -25,42 +25,33 @@ function notUndefined<T>(x: T | undefined): x is T {
...
   public appName: string;
 
   constructor({
-    cwd,
+    cwd = process.cwd(),
     serverFilePath,
-    socketServer,
-    suricate,
     appName,
+    nodeEnv = {
```
Comment (ranyitz) [inline packages/yoshi-common/src/server-process.ts]: I think that this variable name would be better named `env` => all of the environment parameters for the process, in oppose to `nodeEnv` which has a disambiguation with `NODE_ENV`

File: packages/yoshi-flow-app/src/scripts/serve.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+      );
+      return;
+    }
+
+    const serverProcess = new ServerProcess({
+      serverFilePath,
+      appName: config.name,
+    });
+
+    const servers = await Promise.all([
```
Comment (ranyitz) [inline packages/yoshi-flow-app/src/scripts/serve.ts]:
```suggestion
    const [, cdn] = await Promise.all([
```

Would be a bit nicer

File: test/typescript/features/serve/https/serve.test.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  templateDir: __dirname,
+  projectType: 'typescript',
+});
+
+describe('Yoshi serve https', () => {
+  it('should return prod build when running yoshi serve', async () => {
+    await scripts.build();
+
+    await scripts.serve(async () => {
+      const browser = await launch({ ignoreHTTPSErrors: true });
```
Comment (ranyitz) [inline test/typescript/features/serve/https/serve.test.ts]: Awesome

File: packages/yoshi-common/src/server-process.ts (hunk 9 above)
Comment (Gongreg) [reply packages/yoshi-common/src/server-process.ts]: Good point. Will fix it.

File: packages/yoshi-flow-app/src/scripts/serve.ts (hunk 10 above)
Comment (Gongreg) [reply packages/yoshi-flow-app/src/scripts/serve.ts]: ah, I am used to const [_, cdn] = await Promise.all([ and it didn't like it. Thanks

**Comment by Gongreg** [comment]
@ranyitz Fixed up the final comments.

Regarding the documentation maybe we should create a separate section in yoshi docs? It depends if we are going to have more node API exposed in the future.

**Pull request merged**